File: python/techai_webutils/src/techai_webutils/core/interfaces/retrier.py
"""Retrier interface for automatic retry on transient failures.

Mirrors Go's ``interfaces.Retrier``.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable


class Retrier(ABC):
    """Executes operations with automatic retry on transient failures."""

    @abstractmethod
    async def retry(self, op: Callable[[], Awaitable[object]]) -> object:
        """Execute op with automatic retry on transient failures.

        Returns the result on success, or raises the last error after all
        retries are exhausted.
        """
        ...

File: go/core/interfaces/retrier.go
package interfaces

import "context"

// Retrier executes operations with automatic retry on transient failures.
//
// Implementations: exponential backoff with jitter (default).
//
// All service code depends on this interface, never on a concrete retry
// library. Swap implementations via the retry.New factory.
type Retrier interface {
	// Retry executes op with automatic retry on transient failures.
	// Returns nil on success, or the last error after all retries are exhausted.
	Retry(ctx context.Context, op func() error) error
}
